File: __Viejo/Formularios/Agregar/add_produccion.vb
Public Class add_produccion

    Private recibido As Boolean = False
    Private idUsuario As Integer
    Private idUnico As String


    Private Sub add_produccion_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        form = Me

        'Cargo todos los proveedores
        cargar_combo(cmb_proveedor, "SELECT id_proveedor, razon_social FROM proveedores WHERE activo = '1' ORDER BY razon_social ASC", basedb, "razon_social", "id_proveedor")
        cmb_proveedor.SelectedIndex = 0
        cmb_proveedor.Text = "Seleccione un proveedor..."


        If edicion Or borrado = True Then
            lbl_nProduccion.Text = edita_produccion.id_produccion
            lbl_produccion.Visible = True
            lbl_nProduccion.Visible = True
            lbl_fechaCarga.Text = DateTime.Parse(edita_produccion.fecha_carga)

            If edita_produccion.fecha_envio <> "" Then
                lbl_fechaEnvio.Text = DateTime.Parse(edita_produccion.fecha_envio)
                cmd_enviar.Text = "Cerrar pedido"
                dg_viewProduccion.ContextMenuStrip = cms_enviado
                recibido = True
            Else
                lbl_fechaEnvio.Text = ""
            End If

            If edita_produccion.fecha_recepcion <> "" Then
                lbl_fechaRecepcion.Text = DateTime.Parse(edita_produccion.fecha_recepcion)
            Else
                lbl_fechaRecepcion.Text = ""
            End If

            'Dim pv As New proveedor
            'PV = info_proveedor(edita_produccion.id_proveedor)
            cmb_proveedor.SelectedValue = edita_produccion.id_proveedor
            cmb_proveedor.Enabled = False

            actualizarDataGrid()
        Else
            generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
            lbl_fechaCarga.Text = Hoy()
            lbl_fechaEnvio.Text = ""
            lbl_fechaRecepcion.Text = ""
        End If

        If borrado Or (edita_produccion.id_produccion <> 0 And edita_produccion.activo = False) Then
            cmb_proveedor.Enabled = False
            cmd_add_item.Enabled = False
            pic_searchProveedor.Enabled = False
            'dg_viewProduccion.Enabled = False
            chk_imprimir.Enabled = False
            chk_secuencia.Enabled = False
            cmd_ok.Enabled = False
            cmd_enviar.Enabled = False
        End If

        If borrado Then
            If Not edita_produccion.activo Then
                Me.Show()
                MsgBox("No se puede borrar un pedido de producción ya recibido", vbOKOnly + vbExclamation, "Centrex")
                closeandupdate(Me)
                Exit Sub
            End If

            cmd_exit.Enabled = False
            Me.Show()
            If MsgBox("¿Está seguro que desea borrar este pedido de producción?", vbYesNo + vbQuestion) = MsgBoxResult.Yes Then
                'borrartbl("tmpproduccion_items", True)
                generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
                ejecutarSQL("DELETE FROM produccion_items WHERE id_produccion = '" + edita_produccion.id_produccion.ToString + "'")
                If borrarProduccion(edita_produccion) = False Then
                    If (MsgBox("Ocurrió un error al realizar el borrado del pedido de producción, ¿desea intetar desactivarlo para que no aparezca en la búsqueda?" _
                     , MsgBoxStyle.Question + MsgBoxStyle.YesNo)) = vbYes Then
                        'Realizo un borrado lógico
                        If updateProduccion(edita_produccion, True) = True Then
                            MsgBox("Se ha podido realizar un borrado lógico, pero el pedido no se borró definitivamente." + Chr(13) +
                                "Esto posiblemente se deba a que el pedido, tiene operaciones realizadas y por lo tanto no podrá borrarse", vbInformation)
                        Else
                            MsgBox("No se ha podido borrar el pedido de producción.")
                        End If
                    End If
                End If
            End If
            closeandupdate(Me)
        End If
    End Sub

    Private Sub add_produccion_FormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
        id = 0
        Dim prod As New produccion
        edita_produccion = prod
        'borro la tabla temporal
        'borrartbl("tmpproduccion_items", True)
        generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
        'restauro los que se borraron porque no se guardaron los cambios
        activaitems("produccion_items")
        edicion = False
        closeandupdate(Me)
    End Sub

    Private Sub cmd_add_item_Click(sender As Object, e As EventArgs) Handles cmd_add_item.Click
        Dim tmpTabla As String
        Dim tmpActivo As Boolean

        tmpTabla = tabla
        tmpActivo = activo
        tabla = "items_sinDescuento"
        activo = True

        agregaitem = True
        Me.Enabled = False

        Dim srch As New search(True, False, True)
        srch.ShowDialog()
        tabla = tmpTabla
        activo = tmpActivo
        agregaitem = False
        actualizarDataGrid()
        If Tiene_Items_Asociados(id) Then
            Dim frm_detalle_prod As New frm_detalle_asoc_produccion(id)
            frm_detalle_prod.ShowDialog()
        End If
    End Sub

    Private Sub cmd_exit_Click(sender As Object, e As EventArgs) Handles cmd_exit.Click
        closeandupdate(Me)
    End Sub

    Private Sub cmd_ok_Click(sender As Object, e As EventArgs) Handles cmd_ok.Click
        Dim ultimaProd As produccion = Nothing
        Dim p As New produccion
        Dim sqlstr As String = ""

        If cmb_proveedor.Text = "Seleccione un proveedor..." Or cmb_proveedor.Text = "" Then
            MsgBox("El campo 'Proveedor' es obligatorio y está vacio")
            Exit Sub
        ElseIf dg_viewProduccion.Rows.Count = 0 Then
            MsgBox("No hay items cargados")
            Exit Sub
        End If

        p.id_proveedor = cmb_proveedor.SelectedValue
        p.fecha_carga = lbl_fechaCarga.Text
        If lbl_fechaEnvio.Text <> "" Then
            p.fecha_envio = lbl_fechaEnvio.Text
            p.enviado = True
        Else
            p.enviado = False
        End If

        If lbl_fechaRecepcion.Text <> "" Then
            p.fecha_recepcion = lbl_fechaRecepcion.Text
            p.recibido = True
            p.activo = False
        Else
            p.recibido = False
            p.activo = True
        End If

        If edicion Then
            p.id_produccion = edita_produccion.id_produccion
            ultimaProd = p
            If Not updateProduccion(p) Then
                MsgBox("Hubo un problema al actualizar el pedido de producción.", vbExclamation)
                closeandupdate(Me)
            End If
            guardarProduccion(edita_produccion.id_produccion)
            If p.recibido Then
                'Si se recibio la mercadería, igualo los productos recibidos si no es que se explicitó que son distintos
                'y hago lo mismo con las cantidades
                sqlstr = "UPDATE produccion_items " &
                        "SET id_item_recibido = id_item " &
                        "WHERE id_produccion = '" + edita_produccion.id_produccion.ToString + "' " &
                        "AND id_item_recibido IS NULL"
                ejecutarSQL(sqlstr)

                sqlstr = "UPDATE produccion_items " &
                        "SET cantidad_recibida = cantidad " &
                        "WHERE id_produccion = '" + edita_produccion.id_produccion.ToString + "' " &
                        "AND cantidad_recibida IS NULL"
                ejecutarSQL(sqlstr)

                'Cierro el pedido para que se actualice el stock
                sqlstr = "UPDATE produccion_items " &
                        "SET activo = '0' " &
                        "WHERE id_produccion = '" + edita_produccion.id_produccion.ToString + "'"
                ejecutarSQL(sqlstr)
            End If
            'borrartbl("tmpproduccion_items", True)
            generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
        Else
            'Agrego el pedido de producción
            If addProduccion(p) Then
                ultimaProd = info_produccion()
                'Agrego los items al pedido
                If Not guardarProduccion(ultimaProd.id_produccion) Then
                    MsgBox("Hubo un problema al agregar el pedido de producción.", vbExclamation)
                    'borrartbl("tmpproduccion_asocItems")
                    'borrartbl("tmpproduccion_items", True)
                    generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
                    closeandupdate(Me)
                Else
                    'borrartbl("tmpproduccion_asocItems")
                    'borrartbl("tmpproduccion_items", True)
                    generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
                End If
            Else
                MsgBox("Hubo un problema al agregar el pedido de producción.", vbExclamation)
                'borrartbl("tmpproduccion_asocItems")
                'borrartbl("tmpproduccion_items", True)
                generales_multiUsuario.borrarTmpProduccion(usuario_logueado.id_usuario)
                closeandupdate(Me)
            End If
        End If

        If chk_imprimir.Checked Then
            Dim frm As New frm_prnReportes("rpt_produccion", "datos_empresa", "produccion_cabecera", "produccion_detalle", "DS_empresa",
                                            "Produccion_cabecera", "Produccion_detalle", ultimaProd.id_produccion)
            frm.ShowDialog()
        End If

        If chk_secuencia.Checked = True Then
            'borrartbl("tmpproduccion_items",true) 'Ya debería estar borrada la tabla
            'cmb_proveedor.SelectedValue = 0
            'cmb_proveedor.Text = "Seleccione un proveedor..."
            actualizarDataGrid()
        Else
            closeandupdate(Me)
        End If
    End Sub

    Private Sub pic_searchProveedor_Click(sender As Object, e As EventArgs) Handles pic_searchProveedor.Click
        'busqueda
        Dim tmp As String
        tmp = tabla
        tabla = "proveedores"
        Me.Enabled = False
        search.ShowDialog()
        tabla = tmp

        'Establezco la opción del combo, si es 0 elijo el cliente default
        If id = 0 Then id = id_proveedor_default
        updateform(id.ToString, cmb_proveedor)
    End Sub

    Private Sub dg_view_DoubleClick(sender As Object, e As EventArgs) Handles dg_viewProduccion.DoubleClick
        Dim seleccionado As String
        Dim seleccionado_b As String

        If edita_produccion.id_produccion <> 0 And Not edita_produccion.activo Then
            Exit Sub
        ElseIf edicion And edita_produccion.enviado Then
            MsgBox("No puede editar un item ya enviado al proveedor", vbExclamation + vbOKOnly, "Centrex")
            Exit Sub
        End If

        'Obtengo el ID del item
        seleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString()
        seleccionado = Microsoft.VisualBasic.Right(seleccionado, (seleccionado.Length - InStr(seleccionado, "-")))

        edita_item = info_item(seleccionado)

        'Obtengo el ID interno de la tabla tmppedidos_items
        seleccionado_b = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString
        edita_item.id_item_temporal = Microsoft.VisualBasic.Left(seleccionado_b, (InStr(seleccionado_b, "-") - 1))

        Dim agregaItemFrm As New infoagregaitem(True, False, True, idUsuario, idUnico)
        agregaItemFrm.ShowDialog()

        'Dim i As New item
        'edita_item = info_item(id)
        actualizarDataGrid()
        If Tiene_Items_Asociados(edita_item.id_item) Then
            If edita_produccion.id_produccion <> Nothing And edita_produccion.id_produccion <> 0 Then
                'Ya hay un pedido de producción guardado
                Dim frm_detalle_prod As New frm_detalle_asoc_produccion(edita_item.id_item, edita_produccion.id_produccion)
                frm_detalle_prod.ShowDialog()
            Else
                'Se está dando de alta el pedido de producción
                Dim frm_detalle_prod As New frm_detalle_asoc_produccion(edita_item.id_item)
                frm_detalle_prod.ShowDialog()
            End If
        End If
    End Sub

    Private Sub EditarToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles EditarToolStripMenuItem.Click
        dg_view_DoubleClick(Nothing, Nothing)
    End Sub

    Private Sub BorrarToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles BorrarToolStripMenuItem.Click
        Dim seleccionado As String
        Dim id_tmpPedidoItem_seleccionado As String

        If dg_viewProduccion.Rows.Count = 0 Then Exit Sub

        If edicion And edita_produccion.enviado Then
            MsgBox("No puede editar un item ya enviado al proveedor", vbExclamation + vbOKOnly, "Centrex")
            Exit Sub
        End If

        seleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString()

        'Obtengo el ID interno de la tabla tmppedidos_items
        id_tmpPedidoItem_seleccionado = Microsoft.VisualBasic.Left(seleccionado, (InStr(seleccionado, "-") - 1))

        borraritemCargado(id_tmpPedidoItem_seleccionado)
    End Sub

    Private Sub dg_view_MouseDown(sender As Object, e As MouseEventArgs) Handles dg_viewProduccion.MouseDown
        If e.Button = Windows.Forms.MouseButtons.Right Then
            With Me.dg_viewProduccion
                Dim Hitest As DataGridView.HitTestInfo = .HitTest(e.X, e.Y)
                If Hitest.Type = DataGridViewHitTestType.Cell Then
                    .CurrentCell = .Rows(Hitest.RowIndex).Cells(Hitest.ColumnIndex)
                End If
                Me.cms_general.Visible = True
            End With
        End If
    End Sub

    Private Sub cmb_proveedor_KeyPress(sender As Object, e As KeyPressEventArgs) Handles cmb_proveedor.KeyPress
        'e.KeyChar = ""
    End Sub

    Private Sub cmd_enviar_Click(sender As Object, e As EventArgs) Handles cmd_enviar.Click
        If recibido Then
            If MsgBox("ATENCIÓN: Si finaliza la recepción, NO PODRÁ realizar más modificaciones en la orden de producción." & vbCrLf &
                "Se supondrá que los artículos y cantidades que regresaron fueron las mismas que se solicitaron para aquellos en los cuales no haya seleccionado expresamente lo contrario " & vbCrLf &
                "¿Está seguro?", vbYesNo + vbQuestion, "Centrex") = vbYes Then
                lbl_fechaRecepcion.Text = Hoy()
            Else
                Exit Sub
            End If
        Else
            lbl_fechaEnvio.Text = Hoy()
        End If
        cmd_ok_Click(Nothing, Nothing)
    End Sub

    Private Sub actualizarDataGrid()
        Dim sqlstr As String

        sqlstr = "SELECT CONCAT(ti.id_tmpProduccionItem, '-', ti.id_item) AS 'ID', ti.id_produccionItem AS 'id_produccionItem', " &
                    "CASE WHEN i.id_item IS NULL THEN ti.descript ELSE i.descript END AS 'Producto', " &
                    "ti.cantidad AS 'Cant.', " &
                    "CASE WHEN ti.id_item_recibido IS NULL THEN '' ELSE ii.descript END AS 'Producto Recibido', " &
                    "ti.cantidad_recibida AS 'Cantidad Recibida' " &
                   "FROM tmpproduccion_items AS ti " &
                   "LEFT JOIN items AS i ON ti.id_item = i.id_item " &
                   "LEFT JOIN items AS ii ON ti.id_item_recibido = ii.id_item " &
                   "WHERE ti.activo = '1' AND (i.esMarkup = '0' OR ti.id_item IS NULL) " &
                   "ORDER BY id ASC"
        cargar_datagrid(dg_viewProduccion, sqlstr, basedb)
    End Sub

    Private Sub ModificarArtículoRecibidoToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles ModificarArtículoRecibidoToolStripMenuItem.Click
        Dim seleccionado As String
        Dim sqlstr As String
        Dim tmpTabla As String
        Dim tmpActivo As Boolean
        Dim id_item As Integer
        Dim id_item_tmp As Integer

        'Obtengo el ID interno de la tabla tmpproduccion_items
        seleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString
        id_item_tmp = Microsoft.VisualBasic.Left(seleccionado, (InStr(seleccionado, "-") - 1))

        tmpTabla = tabla
        tmpActivo = activo
        tabla = "items_sinDescuento"
        activo = True

        Me.Enabled = False
        Dim srch As New search(True, False, False)
        srch.ShowDialog()
        id_item = id

        sqlstr = "UPDATE tmpproduccion_items SET id_item_recibido = '" + id_item.ToString + "' WHERE id_tmpProduccionItem = '" + id_item_tmp.ToString + "'"
        ejecutarSQL(sqlstr)

        actualizarDataGrid()
    End Sub

    Private Sub ModificarCantidadRecibidaToolStripMenuItem_Click(sender As Object, e As EventArgs) Handles ModificarCantidadRecibidaToolStripMenuItem.Click
        Dim seleccionado As String
        Dim sqlstr As String
        Dim cantidad_recibida As Integer

        'Obtengo el ID del item
        seleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString()
        seleccionado = Microsoft.VisualBasic.Right(seleccionado, (seleccionado.Length - InStr(seleccionado, "-")))

        edita_item = info_item(seleccionado)

        'Obtengo el ID interno de la tabla tmpproduccion_items
        seleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString
        edita_item.id_item_temporal = Microsoft.VisualBasic.Left(seleccionado, (InStr(seleccionado, "-") - 1))

        Dim agregaItem As New infoagregaitem(True, False, False, idUsuario, idUnico)
        agregaItem.ShowDialog()
        cantidad_recibida = agregaItem.cant

        sqlstr = "UPDATE tmpproduccion_items SET cantidad_recibida = '" + cantidad_recibida.ToString + "' WHERE id_tmpProduccionItem = '" + edita_item.id_item_temporal.ToString + "'"
        ejecutarSQL(sqlstr)

        actualizarDataGrid()
    End Sub

    'Private Sub dg_viewProduccion_CellMouseDoubleClick(sender As Object, e As DataGridViewCellMouseEventArgs) Handles dg_viewProduccion.CellMouseDoubleClick
    '    Dim idItemSeleccionado As String

    '    If dg_viewProduccion.Rows.Count = 0 Then Exit Sub

    '    idItemSeleccionado = dg_viewProduccion.CurrentRow.Cells(0).Value.ToString

    '    idItemSeleccionado = Strings.Right(idItemSeleccionado, (InStr(idItemSeleccionado, "-") - 1))

    '    If Tiene_Items_Asociados(idItemSeleccionado) Then
    '        'Dim frm_detalle_prod As New frm_detalle_asoc_produccion(id)
    '        'frm_detalle_prod.ShowDialog()
    '        MsgBox("Tiene asociados")
    '    Else
    '        MsgBox("No tiene asociados")
    '    End If
    'End Sub
End Class

File: Formularios/Agregar/add_produccion.cs
using System;
using System.Linq;
using System.Windows.Forms;

namespace Centrex
{
    public partial class add_produccion : Form
    {
        private bool recibido = false;
        private int idUsuario;
        private Guid idUnico;
        private readonly CentrexDbContext ctx = new CentrexDbContext();

        public add_produccion()
        {
            InitializeComponent();
        }

        private void add_produccion_Load(object sender, EventArgs e)
        {
            idUsuario = usuario_logueado?.IdUsuario ?? 0;
            idUnico = Generar_ID_Unico();

            // Cargar combo de proveedores
            generales.Cargar_Combo(
                ref cmb_proveedor,
                "ProveedorEntity",
                "RazonSocial",
                "IdProveedor",
                -1,
                true,
                null,
                null
            );
            cmb_proveedor.Text = "Seleccione un proveedor...";

            if (edicion || borrado)
            {
                lbl_nProduccion.Text = edita_produccion.IdProduccion.ToString();
                lbl_produccion.Visible = true;
                lbl_nProduccion.Visible = true;
                lbl_fechaCarga.Text = edita_produccion.FechaCarga.ToString("dd/MM/yyyy");

                if (edita_produccion.FechaEnvio.HasValue)
                {
                    lbl_fechaEnvio.Text = edita_produccion.FechaEnvio.Value.ToString("dd/MM/yyyy");
                    cmd_enviar.Text = "Cerrar pedido";
                    dg_viewProduccion.ContextMenuStrip = cms_enviado;
                    recibido = true;
                }
                else lbl_fechaEnvio.Text = "";

                if (edita_produccion.FechaRecepcion.HasValue)
                    lbl_fechaRecepcion.Text = edita_produccion.FechaRecepcion.Value.ToString("dd/MM/yyyy");
                else lbl_fechaRecepcion.Text = "";

                cmb_proveedor.SelectedValue = edita_produccion.IdProveedor;
                cmb_proveedor.Enabled = false;

                CargarGrillaProduccion();
            }
            else
            {
                borrarTmpProduccion(idUsuario);
                lbl_fechaCarga.Text = generales.Hoy();
                lbl_fechaEnvio.Text = "";
                lbl_fechaRecepcion.Text = "";
            }

            if (borrado ||
                (edita_produccion.IdProduccion != 0 && !edita_produccion.Activo))
            {
                cmb_proveedor.Enabled = false;
                cmd_add_item.Enabled = false;
                pic_searchProveedor.Enabled = false;
                chk_imprimir.Enabled = false;
                chk_secuencia.Enabled = false;
                cmd_ok.Enabled = false;
                cmd_enviar.Enabled = false;
            }
        }

        private void add_produccion_FormClosed(object sender, FormClosedEventArgs e)
        {
            id = 0;
            edita_produccion = new ProduccionEntity();
            borrarTmpProduccion(idUsuario);
            generales.ActivaItems("produccion_items");
            edicion = false;
            generales.closeandupdate(this);
        }

        private void cmd_add_item_Click(object sender, EventArgs e)
        {
            var tmpTabla = tabla;
            var tmpActivo = activo;
            tabla = "items_sinDescuento";
            activo = true;
            agregaitem = true;

            Enabled = false;
            var srch = new search(true, false, true);
            srch.ShowDialog();
            tabla = tmpTabla;
            activo = tmpActivo;
            agregaitem = false;
            Enabled = true;

            if (id > 0)
            {
                edita_item = mitem.info_item(id);
                var agregaItemFrm = new infoagregaitem(true, false, true, idUsuario, idUnico);
                agregaItemFrm.ShowDialog();
                edita_item = null;
            }

            if (asocitems.Tiene_Items_Asociados(id))
            {
                var frm_detalle_prod = new frm_detalle_asoc_produccion(id);
                frm_detalle_prod.ShowDialog();
            }

            id = 0;
            CargarGrillaProduccion();
        }

        private void cmd_exit_Click(object sender, EventArgs e)
        {
            generales.closeandupdate(this);
        }

        private void cmd_ok_Click(object sender, EventArgs e)
        {
            if (cmb_proveedor.SelectedIndex == -1)
            {
                Interaction.MsgBox("Debe seleccionar un proveedor.", MsgBoxStyle.Exclamation, "Centrex");
                return;
            }

            if (dg_viewProduccion.Rows.Count == 0)
            {
                Interaction.MsgBox("No hay ítems cargados.", MsgBoxStyle.Exclamation, "Centrex");
                return;
            }

            var p = new ProduccionEntity();

            p.IdProveedor = (int)cmb_proveedor.SelectedValue;
            p.FechaCarga = ConversorFechas.GetFecha(lbl_fechaCarga, p.FechaCarga);
            p.Enviado = !string.IsNullOrWhiteSpace(lbl_fechaEnvio.Text);
            p.Recibido = !string.IsNullOrWhiteSpace(lbl_fechaRecepcion.Text);
            p.Activo = string.IsNullOrWhiteSpace(lbl_fechaRecepcion.Text);


            if (!string.IsNullOrWhiteSpace(lbl_fechaEnvio.Text))
                p.FechaEnvio = ConversorFechas.GetFecha(lbl_fechaEnvio, p.FechaEnvio);
            if (!string.IsNullOrWhiteSpace(lbl_fechaRecepcion.Text))
                p.FechaRecepcion = ConversorFechas.GetFecha(lbl_fechaRecepcion, p.FechaRecepcion);

            if (edicion)
            {
                p.IdProduccion = edita_produccion.IdProduccion;
                ctx.Entry(p).State = EntityState.Modified;
            }
            else
            {
                ctx.ProduccionEntity.Add(p);
            }

            ctx.SaveChanges();

            borrarTmpProduccion(idUsuario);

            if (chk_imprimir.Checked)
            {
                //var frm = new frm_prnOrdenProduccion(p.IdProduccion);
                //frm.ShowDialog();
            }

            if (chk_secuencia.Checked)
                CargarGrillaProduccion();
            else
                generales.closeandupdate(this);
        }

        private void cmd_enviar_Click(object sender, EventArgs e)
        {
            if (recibido)
            {
                if (Interaction.MsgBox("¿Finalizar recepción? No se podrá modificar luego.", MsgBoxStyle.YesNo | MsgBoxStyle.Question) == MsgBoxResult.Yes)
                    lbl_fechaRecepcion.Text = generales.Hoy();
                else
                    return;
            }
            else
            {
                lbl_fechaEnvio.Text = generales.Hoy();
            }

            cmd_ok_Click(null, null);
        }

        private void CargarGrillaProduccion()
        {
            try
            {
                var query = ctx.TmpProduccionItemEntity
                    .Include(t => t.IdItemNavigation)
                    .Include(t => t.IdItemRecibidoNavigation)
                    .Where(t => t.Activo == true && t.IdUsuario == idUsuario)
                    .OrderBy(t => t.IdTmpProduccionItem);

                var result = new DataGridQueryResult
                {
                    Query = query,
                    GridName = "produccion",
                    EsMaterializada = false
                };

                LoadDataGridDynamic.LoadDataGridWithEntityAsync(dg_viewProduccion, result).GetAwaiter().GetResult();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al cargar grilla: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void EditarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            try
            {
                if (dg_viewProduccion.CurrentRow == null)
                    return;

                // Simula doble click (igual al VB original)
                //dg_viewProduccion_DoubleClick(null, null);
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al editar el ítem: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void BorrarToolStripMenuItem_Click(object sender, EventArgs e)
        {
            try
            {
                if (dg_viewProduccion.CurrentRow == null)
                    return;

                var idTmp = dg_viewProduccion.CurrentRow.Cells[0].Value?.ToString();
                if (string.IsNullOrWhiteSpace(idTmp))
                    return;

                // Extrae el ID temporal antes del guion
                var idTmpProduccionItem = idTmp.Split('-')[0];

                if (Interaction.MsgBox("¿Seguro que desea eliminar el ítem seleccionado?",
                                       MsgBoxStyle.YesNo | MsgBoxStyle.Question, "Centrex") != MsgBoxResult.Yes)
                    return;

                generales.ejecutarSQL($"DELETE FROM tmpproduccion_items WHERE id_tmpProduccionItem = '{idTmpProduccionItem}'");

                CargarGrillaProduccion();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al borrar ítem: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void dg_view_MouseDown(object sender, MouseEventArgs e)
        {
            try
            {
                if (e.Button == MouseButtons.Right)
                {
                    var hit = dg_viewProduccion.HitTest(e.X, e.Y);
                    if (hit.Type == DataGridViewHitTestType.Cell)
                    {
                        dg_viewProduccion.CurrentCell = dg_viewProduccion.Rows[hit.RowIndex].Cells[hit.ColumnIndex];
                        cms_general.Show(dg_viewProduccion, e.Location);
                    }
                }
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al abrir menú contextual: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void ModificarArtículoRecibidoToolStripMenuItem_Click(object sender, EventArgs e)
        {
            try
            {
                if (dg_viewProduccion.CurrentRow == null) return;

                var seleccionado = dg_viewProduccion.CurrentRow.Cells[0].Value?.ToString();
                if (string.IsNullOrWhiteSpace(seleccionado)) return;

                int idTmpItem = int.Parse(seleccionado.Split('-')[0]);

                tabla = "items_sinDescuento";
                activo = true;
                Enabled = false;

                var srch = new search(true, false, false);
                srch.ShowDialog();
                Enabled = true;

                int idItemNuevo = id;
                if (idItemNuevo <= 0) return;

                generales.ejecutarSQL($"UPDATE tmpproduccion_items SET id_item_recibido = '{idItemNuevo}' WHERE id_tmpProduccionItem = '{idTmpItem}'");
                CargarGrillaProduccion();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al modificar artículo recibido: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }

        private void ModificarCantidadRecibidaToolStripMenuItem_Click(object sender, EventArgs e)
        {
            try
            {
                if (dg_viewProduccion.CurrentRow == null) return;

                var seleccionado = dg_viewProduccion.CurrentRow.Cells[0].Value?.ToString();
                if (string.IsNullOrWhiteSpace(seleccionado)) return;

                int idTmpItem = int.Parse(seleccionado.Split('-')[0]);

                var agregaFrm = new infoagregaitem(true, false, false, idUsuario, idUnico);
                agregaFrm.ShowDialog();

                int nuevaCantidad = agregaFrm.cant;
                if (nuevaCantidad < 0) return;

                generales.ejecutarSQL($"UPDATE tmpproduccion_items SET cantidad_recibida = '{nuevaCantidad}' WHERE id_tmpProduccionItem = '{idTmpItem}'");
                CargarGrillaProduccion();
            }
            catch (Exception ex)
            {
                Interaction.MsgBox($"Error al modificar cantidad recibida: {ex.Message}", MsgBoxStyle.Critical, "Centrex");
            }
        }
    }
}
